### PR TITLE
Feature: Pre-commit R

### PR DESCRIPTION
**Pull request by avisionh**
# Summary
This branch adds R pre-commit hooks to the repo so data science projects that use Python **and** R will have pre-commit checks conducted on them.

# Check
- [ ] The R pre-commit hooks work
    + When you don't have `devtools` already installed
    + When you do have `devtools` already installed

A sample file. courtesy of @nacnudus, that triggers the hooks can be found below:

```r
# 3c266dcfef0f244d90685e45f9c1a6d91c4055be4c48c84161ffe9b7936ddc01 # Detect secrets
x = 2 # style-files
# trailing whitespace 
x
"/absolute/path" # lintr
library(styler # parsable-R
browser() # no-browser-statement
# No newline character at end of file
```

## Note
It's quite tricky using the `.Rprofile` to install packages, including those specified in the `DESCRIPTION` file upon loading R. The best solution was found [here](https://stackoverflow.com/a/55539569/13416265) which was especially useful in avoiding the invoking of infinite loops...

If you dislike the `.Rprofile` and `startup.R` method, there are two alternatives:

1. Request the user enters the below in their terminal before running any R code. This may require the user to set the path to R.
```r
R
install.packages(pkgs = 'devtools', Ncpus = 1)
devtools::install()
```
2.  Utilise `renv`. This requires the installation of many other packages so isn't exactly lightweight. However, abstracts the details of the installation of these pre-commit hooks well as the packages are downloaded and installed before the user starts R, plus it is throroughly tested.

## Previous work
This branch builds on the work done in #9, but simplifies it by just focussing on the pre-commit hooks. In #9, there was the continual risk that using `renv` would complicate the branch too much. Indeed, it does seem to add a few more packages to the repo.

### CODE REVIEW

**Review by nacnudus — CHANGES_REQUESTED** [review]
This was a very smooth experience once I had done a couple of things not mentioned in the main README at the root of the repository.

```sh
git init
make requirements
```

They had to be done in that order, otherwise `make requirements` fails eventually.

After that, I created and staged a file `src/utils/bad.R` from your suggestion in this PR.  All bad things were detected.  I corrected the ones that weren't merely stylistic, staged the file, and reran.  The stylistic faults were corrected by the {styler} package, with an informative message, but the code was still not committed.  Finally, I staged the latest changes, and committed without further trouble.  This is all as I expected.

File: {{ cookiecutter.repo_name }}/.lintr
Comment (avisionh) [reply {{ cookiecutter.repo_name }}/.lintr]: Named the argument now in https://github.com/ukgovdatascience/govcookiecutter/pull/16/commits/e680f2259cc3066e97400da1c56c5861da8564d2. On the tidyverse style guide recommendation, I extended it to 120 to be consistent with the `flake8` one specified in the file below:
https://github.com/ukgovdatascience/govcookiecutter/blob/master/.flake8#L8

File: {{ cookiecutter.repo_name }}/.pre-commit-config.yaml
Comment (avisionh) [reply {{ cookiecutter.repo_name }}/.pre-commit-config.yaml]: Good point, done! https://github.com/ukgovdatascience/govcookiecutter/pull/16/commits/acc86a0e97a9142fb6f2dcfd8f9bda479afd38ae

**Comment by avisionh** [comment]
> This was a very smooth experience once I had done a couple of things not mentioned in the main README at the root of the repository.
> 
> ```shell
> git init
> make requirements
> ```
> 
> They had to be done in that order, otherwise `make requirements` fails eventually.
> 
> After that, I created and staged a file `src/utils/bad.R` from your suggestion in this PR. All bad things were detected. I corrected the ones that weren't merely stylistic, staged the file, and reran. The stylistic faults were corrected by the {styler} package, with an informative message, but the code was still not committed. Finally, I staged the latest changes, and committed without further trouble. This is all as I expected.

Thanks for the review @nacnudus! On updating the `README.md`, I think I'll leave that to @ESKYoung as he has an idea on what he wants to do with it. Hope that's okay.

File: {{ cookiecutter.repo_name }}/.pre-commit-config.yaml
Comment (exfalsoquodlibet) [inline {{ cookiecutter.repo_name }}/.pre-commit-config.yaml:54]: Do we really want to be so strict to include by default data files `.RData`, `.Rds` and `.rds`? If so, `.rda` is missing from the list.

However, I do not think we want this. Users may want to commit data or trained models without having to modify the `.pre-commit-config.yaml` (and can exclude them from committing these files via `.gitignore` as usual).

Through the python pre-commit, we check for any large files (over 5 MB) being committed (which will cover both python- and r- generated objects). I think that's enough? 

@nacnudus @avisionh

**Review by exfalsoquodlibet — COMMENTED** [review]
Thanks @avisionh! 

All hooks triggered 🎉 

Just a comment on the pre-commit hook for R data/object files - I do not think we want to be so strict :)

**Some more general considerations.**

I think this is an option for users who need/want to use both Python and R for a project. In this "multi-language framework" sense, it is a nice option to have and it is good that it is indeed optional. For this reason, and phrased in these terms, I'd be happy for this to be added to `govcookiecutter`.

I still have concerns that someone working on a purely R project will go through the "troubles" of (especially if they are not also python users):
- installing python
- create a python virtual environment 😱
- ending up with a project structure that does not usually reflect most R project structures/packages and is full of unnecessary python stuff.

Also, important to flag in the README or somewhere else that a python virtual environment and the pythonic `make requirements` is also a necessary step, even if one only wants to use R for their project.

If we think about why a user working on a purely R project may want to use `govcookiecutter` is probably to have access to the following functionalities:
- pre-commit hooks (especially the secret file ones)
- AQA docs and nudges (e.g., PR Template)
There is a recent package (by the same author our R hooks have been taken): https://cran.r-project.org/web/packages/precommit/index.html
which makes me wonder whether a purely-R cookiecutter could be possible (as an option).

Not saying we shall be doing it now,  it may be worth checking what the appetite for both options (Python+R, separate python and r options) are at the show & tell Eric is organising. @ESKYoung @nacnudus

**Comment by avisionh** [comment]
> Thanks @avisionh!
> 
> All hooks triggered 🎉
> 
> Just a comment on the pre-commit hook for R data/object files - I do not think we want to be so strict :)
> 
> **Some more general considerations.**
> 
> ...

Thanks for the thoughtful feedback @exfalsoquodlibet! Yes, I agree, I really do think we can add more value by having an R-only version of govcookiecutter and as you highlighted, it is possible to have one because of the [precommit](https://github.com/lorenzwalthert/precommit) package. This package still uses the pre-commit framework though but just abstracts from the details of creating a virtual environment. In this way, it actually installs the pre-commit framework on your system.

File: {{ cookiecutter.repo_name }}/.pre-commit-config.yaml
Comment (nacnudus) [reply {{ cookiecutter.repo_name }}/.pre-commit-config.yaml:54]: [Hooks can be skipped](https://pre-commit.com/#temporarily-disabling-hooks) with `SKIP=flake8 git commit -m "foo" (if using git at the command line), but I agree that this could be handled by `.gitignore`.

Comment (avisionh) [reply {{ cookiecutter.repo_name }}/.pre-commit-config.yaml]: These are very good points. Added it to https://github.com/ukgovdatascience/govcookiecutter/pull/16/commits/e13ad5297ebdbfca6f7237c4783fd12a2781acf2 now!

File: {{ cookiecutter.repo_name }}/DESCRIPTION
Comment (ESKYoung) [inline {{ cookiecutter.repo_name }}/DESCRIPTION:11]: Would strongly recommend using Jinja2 templating to automatically fill these values that are supplied by `cookiecutter.json` or pre-defined.  Not sure how to define an organisation, so followed this [SO answer](https://stackoverflow.com/a/50688358).

```suggestion
Package: {{ cookiecutter.repo_name }}
Title: {{ cookiecutter.project_name }}
Version: {{ cookiecutter.project_version }}
Authors@R:
    person(given = "{{ cookiecutter.organisation_name }}",
           role = c("aut", "cre"),
           email = "{{ cookiecutter.contact_email }}")
Description: {{ cookiecutter.overview }}
License: Unless stated otherwise, the codebase is released under 
    the MIT License. This covers both the codebase and any sample 
    code in the documentation. The documentation is © Crown 
    copyright and available under the terms of the Open Government 
    3.0 licence.
```

**Review by ESKYoung — CHANGES_REQUESTED** [review]
Thanks for the contributions @avisionh! Looks good - won't comment on the tripping of R hooks as both are/were better covered by @nacnudus and @exfalsoquodlibet.

One requested change with you current code, but also noting the following should be added:

- Missing documentation on file descriptions in `docs/structure/README.md` (with associated Jinja2 conditional statements)
- R files (`.Rprofile`, `.lintr`, `DESCRIPTION`, `startup.R`) exist even for a pure Python project.

For the latter, would suggest modifying the function from [here](https://github.com/ukgovdatascience/govcookiecutter/blob/0043d68585cbfc70b87b4a2792dc9e80d6544de9/hooks/post_gen_project.py#L83) (and its associated test [here](https://github.com/ukgovdatascience/govcookiecutter/blob/add_conda_support/tests/test_post_gen_project/test_set_requirements_or_environment_file.py)) to remove these files if R hooks are not selected.

Understand this is a bit of a patchwork, and it would be better to have a file defining all the R files for deletion, but this can be added at a later date. Happy to assist with amending the above code (since I wrote it!), and/or help clarify cookiecutter hooks.

**Comment by avisionh** [comment]
Thanks @ESKYoung for the helpful review! I've implemented all your suggestions, though tweaking the MIT license part of the `DESCRIPTION` file as per R best practice and @nacnudus's comments.

I'm happy to implement the conditional removal/inclusion of R files once # is merged in. Want to avoid causing conflicts!

**Review by ESKYoung — COMMENTED** [review]
This all looks fine on my end, great work @avisionh! I'll approve once the demo next week has been presented!

**Review by ESKYoung — APPROVED** [review]

**Review by nacnudus — APPROVED** [review]

**Pull request merged**
